haskell-compile.el

```diff
@@ -143,19 +158,50 @@ node `(haskell-mode)compilation' for more details."
   (interactive "P")
   (save-some-buffers (not compilation-ask-about-save)
                          compilation-save-buffers-predicate)
-  (let* ((cabdir (haskell-cabal-find-dir))
-         (command1 (if (eq edit-command '-)
-                       haskell-compile-cabal-build-alt-command
-                     haskell-compile-cabal-build-command))
-         (srcname (buffer-file-name))
-         (command (if cabdir
-                      (format command1 cabdir)
-                    (if (and srcname (derived-mode-p 'haskell-mode))
-                        (format haskell-compile-command srcname)
-                      command1))))
-    (when (and edit-command (not (eq edit-command '-)))
-      (setq command (compilation-read-command command)))
+  (if-let ((stackdir (locate-dominating-file buffer-file-name "stack.yaml")))
+      (haskell-compile-stack stackdir edit-command)
+    (if-let ((cabaldir (haskell-cabal-find-dir)))
+        (haskell-compile-cabal cabaldir edit-command)
+      (let ((srcfile (buffer-file-name)))
+        (haskell-compile-ghc srcfile edit-command)))))
+
+(setq haskell-compile-stack-last nil)
+(defun haskell-compile-stack (dir edit-command)
+  (let* ((default (or haskell-compile-stack-last
+                      haskell-compile-stack-build-command))
+         (template (pcase edit-command
+                     ('nil default)
+                     ('-  haskell-compile-stack-build-alt-command)
+                     (_   (compilation-read-command default))))
+         (command (format template dir)))
+    (unless (eq edit-command '-)
+      (setq haskell-compile-stack-last template))
+    (compilation-start command 'haskell-compilation-mode)))
+
+(setq haskell-compile-cabal-last nil)
+(defun haskell-compile-cabal (dir edit-command)
```

Comment:
I see nothing wrong with using macros here for `ghc/cabal/stack`.

Reply:
I think I'd like to add a lot more functionality to this procedure, eventually (e.g. test/run/other support). Making it a macro will introduce tech debt that may be hard to remove. I'll look into passing the symbol of the `-last` to see if that's a possible way to update the state.